state-sync/aptos-data-client/src/latency_monitor.rs

```diff
@@ -0,0 +1,495 @@
+                        warn!(
+                            (LogSchema::new(LogEntry::LatencyMonitor)
+                                .event(LogEvent::StorageReadFailed)
+                                .message(&format!("Unable to read the highest synced version: {:?}", error)))
+                        );
+                    );
+                    continue; // Continue to the next round
+                },
+            };
+
+            // Update the latency metrics for all versions that we've now synced
+            self.update_latency_metrics(highest_synced_version);
+
+            // Get the highest advertised version from the global data summary
+            let advertised_data = &self.data_client.get_global_data_summary().advertised_data;
+            let highest_advertised_version = match advertised_data.highest_synced_ledger_info() {
+                Some(ledger_info) => ledger_info.ledger_info().version(),
+                None => {
+                    sample!(
+                        SampleRate::Duration(Duration::from_secs(LATENCY_MONITOR_LOG_FREQ_SECS)),
+                        warn!(
+                            (LogSchema::new(LogEntry::LatencyMonitor)
+                                .event(LogEvent::AggregateSummary)
+                                .message("Unable to get the highest advertised version!"))
+                        );
+                    );
+                    continue; // Continue to the next round
+                },
+            };
+
+            // Update the advertised version timestamps
+            self.update_advertised_version_timestamps(
+                highest_synced_version,
+                highest_advertised_version,
+            );
+        }
+    }
+
+    /// Updates the latency metrics for all versions that have now been synced
+    fn update_latency_metrics(&mut self, highest_synced_version: u64) {
+        // Split the advertised versions into synced and unsynced versions
+        let unsynced_advertised_versions = self
+            .advertised_version_timestamps
+            .split_off(&(highest_synced_version + 1));
+
+        // Update the metrics for all synced versions
+        for (synced_version, (seen_time, seen_timestamp_usecs)) in
+            self.advertised_version_timestamps.iter()
+        {
+            // Update the seen to synced latencies
+            let duration_from_seen_to_synced = self.time_service.now().duration_since(*seen_time);
+            metrics::observe_value_with_label(
+                &metrics::SYNC_LATENCIES,
+                metrics::SEEN_TO_SYNC_LATENCY_LABEL,
+                duration_from_seen_to_synced.as_secs_f64(),
+            );
+
+            // Update the proposal latencies
+            if let Ok(block_timestamp_usecs) = self.storage.get_block_timestamp(*synced_version) {
+                // Update the propose to seen latencies
+                if let Some(duration_from_propose_to_seen) =
```

Comment:
What happens here if we get Err / None do we just drop things?

Reply:
Nothing should be dropped here 😄 If we don't hit these, we won't update the metrics, which I think is the safest behaviour.